File: Strings/valid_pallindrome2.py
// Check whether a string can be a valid palindrome by removing at most one character from it
'''
    Explanation:
    The function valid_palindrome takes a string s as input and returns a boolean indicating whether the 
    string can be a valid palindrome by removing at most one character from it.

    To check if the string can be a valid palindrome, we start by iterating over the string from both the 
    left and right ends at the same time. If we encounter two characters that are not the same, we need to 
    check if removing one of them would make the string a palindrome. We try removing the left character 
    first and create two new substrings - one with the left character removed and one with the right character removed.
    We then check if either of the substrings is a palindrome. If either of the substrings is a palindrome, 
    then the original string can be a palindrome by removing the corresponding character. If neither of 
    the substrings is a palindrome, then the original string cannot be a palindrome even by removing a character.

    If we have checked the whole string and haven't returned yet, it means that the original string is a palindrome.

    The time complexity of the given function is O(n), where n is the length of the input string. 
    This is because we are iterating through the string only once and performing constant-time operations.

    The space complexity of the function is also O(n), where n is the length of the input string. 
    This is because we are using a constant amount of extra space to keep track of the counts of characters. 
    The size of this extra space does not depend on the length of the input string.
'''
# using two-pointer Technique
def valid_palindrome(s: str) -> bool:
    left, right = 0, len(s) - 1
    
    # iterate over the string from left and right at the same time
    while left < right:
        # if the characters at the left and right indices are not the same
        if s[left] != s[right]:
            # we need to check if removing one of them makes the string a palindrome
            # we try removing the left character first
            left_removed = s[left:right]
            right_removed = s[left+1:right+1]
            # if either of the substrings after removal is a palindrome, then the original string is also a palindrome
            return left_removed == left_removed[::-1] or right_removed == right_removed[::-1]
        
        # move to the next pair of characters
        left += 1
        right -= 1
    
    # if we have checked the whole string and haven't returned yet, it's a palindrome
    return True

File: Strings/valid_pallindrome2.java
// Check whether a string can be a valid palindrome by removing at most one character from it
/*
   The given Go program takes a string as input and checks if it can be a valid palindrome by removing at most one
   character from it. The program starts by defining two pointers, left and right, that move towards each other from
   opposite ends of the input string. At each step, if the characters at the left and right pointers are equal,
   then they move towards the center of the string. If the characters are not equal, then we check whether removing
   the character at the left or right pointer would make the rest of the string a palindrome. If both removals fail,
   then the string cannot be a valid palindrome by removing at most one character. The program returns true
   if the string can be made a valid palindrome by removing at most one character, false otherwise.

  The time complexity of this algorithm is O(n), where n is the length of the input string.
  The space complexity is O(1), as we only use constant extra space to store a few variables.

*/
public static boolean isPalindromeWithRemoval(String s) {
  // Define two pointers to check the string from both ends
  int i = 0, j = s.length() - 1;

  while (i < j) {
    // If characters at the pointers are not equal
    if (s.charAt(i) != s.charAt(j)) {
      // Check if removing a character from the left pointer makes it a palindrome
      String leftStr = s.substring(0, i) + s.substring(i + 1);
      if (isPalindrome(leftStr)) {
        return true;
      }
      // Check if removing a character from the right pointer makes it a palindrome
      String rightStr = s.substring(0, j) + s.substring(j + 1);
      if (isPalindrome(rightStr)) {
        return true;
      }
      // If neither option works, it cannot be converted into a palindrome by removing at most one character
      return false;
    }
    // Move the pointers towards the middle of the string
    i++;
    j--;
  }
  return true;
}

public static boolean isPalindrome(String s) {
  // Define two pointers to check the string from both ends
  int i = 0, j = s.length() - 1;

  while (i < j) {
    // If characters at the pointers are not equal, it is not a palindrome
    if (s.charAt(i) != s.charAt(j)) {
      return false;
    }
    // Move the pointers towards the middle of the string
    i++;
    j--;
  }
  return true;
}
